hook-common/src/pgqueue.rs

```diff
@@ -0,0 +1,215 @@
+/// Failed: A job that was unsuccessfully completed by a worker.
+/// Running: A job that was picked up by a worker and it's currentlly being run.
+#[derive(Debug, PartialEq, sqlx::Type)]
+#[sqlx(type_name = "job_status")]
+#[sqlx(rename_all = "lowercase")]
+pub enum JobStatus {
+    Available,
+    Completed,
+    Failed,
+    Running,
+}
+
+/// Allow casting JobStatus from strings.
+impl FromStr for JobStatus {
+    type Err = PgQueueError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "available" => Ok(JobStatus::Available),
+            "completed" => Ok(JobStatus::Completed),
+            "failed" => Ok(JobStatus::Failed),
+            "running" => Ok(JobStatus::Running),
+            invalid => Err(PgQueueError::ParseJobStatusError(invalid.to_owned())),
+        }
+    }
+}
+
+/// JobParameters are stored and read to and from a JSONB field, so we accept anything that fits `sqlx::types::Json`.
+pub type JobParameters<J> = sqlx::types::Json<J>;
+
+/// A Job to be executed by a worker dequeueing a PgQueue.
+#[derive(sqlx::FromRow)]
+pub struct Job<J> {
+    pub id: i64,
+    pub attempt: i32,
+    pub finished_at: Option<DateTime<Utc>>,
+    pub created_at: DateTime<Utc>,
+    pub started_at: Option<DateTime<Utc>>,
+    pub status: JobStatus,
+    pub parameters: sqlx::types::Json<J>,
+}
+
+/// A NewJob to be enqueued into a PgQueue.
+pub struct NewJob<J> {
+    pub attempt: i32,
+    pub finished_at: Option<DateTime<Utc>>,
+    pub started_at: Option<DateTime<Utc>>,
+    pub status: JobStatus,
+    pub parameters: sqlx::types::Json<J>,
+}
+
+impl<J> NewJob<J> {
+    pub fn new(parameters: J) -> Self {
+        Self {
+            attempt: 0,
+            parameters:  sqlx::types::Json(parameters),
+            finished_at: None,
+            started_at: None,
+            status: JobStatus::Available,
+        }
+    }
+}
+
+/// A queue implemented on top of a PostgreSQL table.
+pub struct PgQueue {
+    table: String,
+    pool: PgPool,
+}
+
+pub type PgQueueResult<T> = std::result::Result<T, PgQueueError>;
+
+impl PgQueue {
+    /// Initialize a new PgQueue backed by table in PostgreSQL.
+    pub async fn new(table: &str, url: &str) -> PgQueueResult<Self> {
+        let table = table.to_owned();
+        let pool = PgPoolOptions::new()
+            .connect(url)
+            .await
+            .map_err(|error| PgQueueError::ConnectionError {error})?;
+
+        Ok(Self {
+            table,
+            pool,
+        })
+    }
+
+    /// Dequeue a Job from this PgQueue.
+    pub async fn dequeue<J: DeserializeOwned + std::marker::Send + std::marker::Unpin + 'static>(&self) -> PgQueueResult<Job<J>> {
+        let base_query = format!(
+            r#"
+WITH available_in_queue AS (
+    SELECT
+        id
+    FROM
+        "{0}"
+    WHERE
+        status = 'available'
+    ORDER BY
+        id
+    LIMIT 1
+    FOR UPDATE SKIP LOCKED
```

Comment:
For future readers' sake can you link to https://www.2ndquadrant.com/en/blog/what-is-select-skip-locked-for-in-postgresql-9-5/ (or anything else if you have something better)